tests/interlocked_win32_ut/interlocked_win32_ut.c

```diff
@@ -0,0 +1,708 @@
+MOCKABLE_FUNCTION(, LONG64, mock_InterlockedDecrement64, volatile LONG64*, addend);
+MOCKABLE_FUNCTION(, LONG, mock_InterlockedExchange, volatile LONG*, target, LONG, value);
+MOCKABLE_FUNCTION(, SHORT, mock_InterlockedExchange16, volatile SHORT*, target, SHORT, value);
+MOCKABLE_FUNCTION(, LONG64, mock_InterlockedExchange64, volatile LONG64*, target, LONG64, value);
+MOCKABLE_FUNCTION(, char, mock_InterlockedExchange8, volatile char*, target, char, value);
+MOCKABLE_FUNCTION(, LONG, mock_InterlockedExchangeAdd, volatile LONG*, addend, LONG, value);
+MOCKABLE_FUNCTION(, LONG64, mock_InterlockedExchangeAdd64, volatile LONG64*, addend, LONG64, value);
+MOCKABLE_FUNCTION(, void*, mock_InterlockedExchangePointer, void* volatile*, target, void*, value);
+MOCKABLE_FUNCTION(, LONG, mock_InterlockedIncrement, volatile LONG*, addend);
+MOCKABLE_FUNCTION(, SHORT, mock_InterlockedIncrement16, volatile SHORT*, addend);
+MOCKABLE_FUNCTION(, LONG64, mock_InterlockedIncrement64, volatile LONG64*, addend);
+MOCKABLE_FUNCTION(, LONG, mock_InterlockedOr, volatile LONG*, destination, LONG, value);
+MOCKABLE_FUNCTION(, SHORT, mock_InterlockedOr16, volatile SHORT*, destination, SHORT, value);
+MOCKABLE_FUNCTION(, LONG64, mock_InterlockedOr64, volatile LONG64*, destination, LONG64, value);
+MOCKABLE_FUNCTION(, char, mock_InterlockedOr8, volatile char*, destination, char, value);
+MOCKABLE_FUNCTION(, LONG, mock_InterlockedXor, volatile LONG*, destination, LONG, value);
+MOCKABLE_FUNCTION(, SHORT, mock_InterlockedXor16, volatile SHORT*, destination, SHORT, value);
+MOCKABLE_FUNCTION(, LONG64, mock_InterlockedXor64, volatile LONG64*, destination, LONG64, value);
+MOCKABLE_FUNCTION(, char, mock_InterlockedXor8, volatile char*, destination, char, value);
+#undef ENABLE_MOCKS
+
+#ifdef __cplusplus
+}
+#endif
+
+BEGIN_TEST_SUITE(interlocked_win32_unittests)
+
+TEST_SUITE_INITIALIZE(suite_init)
+{
+    g_testByTest = TEST_MUTEX_CREATE();
+    ASSERT_IS_NOT_NULL(g_testByTest);
+    ASSERT_ARE_EQUAL(int, 0, umock_c_init(on_umock_c_error));
+    ASSERT_ARE_EQUAL(int, 0, umocktypes_windows_register_types());
+    REGISTER_UMOCK_ALIAS_TYPE(SHORT, int16_t);
+
+}
+
+TEST_SUITE_CLEANUP(TestClassCleanup)
+{
+    umock_c_deinit();
+
+    TEST_MUTEX_DESTROY(g_testByTest);
+}
+
+TEST_FUNCTION_INITIALIZE(f)
+{
+    if (TEST_MUTEX_ACQUIRE(g_testByTest))
+    {
+        ASSERT_FAIL("our mutex is ABANDONED. Failure in test framework");
+    }
+
+    umock_c_reset_all_calls();
+}
+
+TEST_FUNCTION_CLEANUP(cleans)
+{
+    TEST_MUTEX_RELEASE(g_testByTest);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_001: [interlocked_add shall call InterlockedAdd from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_002 : [interlocked_add shall return the result of the addition.]*/
+TEST_FUNCTION(interlocked_add_calls_InterlockedAdd)
+{
+    ///arrange
+    volatile int32_t addend = INT32_MAX;
+    int32_t value = INT32_MIN;
+    STRICT_EXPECTED_CALL(mock_InterlockedAdd((volatile LONG*)&addend, (LONG)value))
```

Comment:
>(volatile LONG*) [](start = 45, length = 16)

I am only going to ask this (the second time). why is the cast needed?

My view is simple: either the cast does something to the value, in which case the code will be wrong, as it would not match the definiton domains for LONG/int32_t 

OR

the cast does nothing, in which case why is it needed?

Reply:
If I don't use the cast, I get this error: Warning C4057 'function': 'volatile LONG *' differs in indirection to slightly different base types from 'volatile int32_t *'.

---
In reply to: [427740898](https://github.com/Azure/azure-c-util/pull/43#discussion_r427740898) [](ancestors = 427740898)